File: src/main/java/org/javadsa/Sorting.java
package org.javadsa;
import java.util.Arrays;

public class Sorting {

    public static void bubbleSort(int[] x){
        int i=0;
        boolean swapped = true;

        while (swapped && i<x.length-1){
            swapped = false;
            for (int j=0;j<x.length-1-i;i++){
                if (x[j]>x[j+1]){
                    int temp=x[j];
                    x[j]=x[j+1];
                    x[j+1]=temp;
                }
                swapped = true;
            }
        }
    }

    public static int[] mergeSort(int[] arr) {

        if (arr.length <= 1) {
            return arr;
        } else {
            int mid = arr.length / 2;
            int[] left = mergeSort(Arrays.copyOfRange(arr, 0, mid));
            int[] right = mergeSort(Arrays.copyOfRange(arr, mid, arr.length));

            return merge(left, right);

        }
    }

        private static int[] merge(int[] left, int[] right) {

            int[] result = new int[left.length + right.length];
            int i= 0;
            int j= 0;
            int k= 0;

            while(i<left.length && j<right.length){
                if (left[i] <= right[j]) {
                    result[k] = left[i];
                    i++;
                } else {
                    result[k] = right[j];
                    j++;
                }

                k++;
            }
           // Remaining elements from left
            while (i < left.length) {
                result[k] = left[i];
                i++;
                k++;
            }

            // Remaining elements from right
            while (j < right.length) {
                result[k] = right[j];
                j++;
                k++;
            }

        return result;
    }


        public static void main(String[] args) {

            int[] x = {5, 3, 8, 4, 2};

            int[] result = mergeSort(x);

            System.out.println(Arrays.toString(result));
        }
    }

File: python-dsa/src/dsa/sorting.py
def bubble_sort(x):
    i=0
    swapped=True
    while swapped and i < len(x)-1:
        swapped=False
        for j in range(0,len(x)-i-1):
            if x[j]>x[j+1]:
                temp=x[j]
                x[j]=x[j+1]
                x[j+1]=temp
                swapped=True
        i+=1

def merge_sort(arr):
    if len(arr)<=1:
        return arr
    mid = len(arr) // 2

    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left,right)

def merge(left,right):
    array=[]
    i=j=0
    while i<len(left) and j<len(right):
        if left[i]<=right[j]:
            array.append(left[i])
            i+=1
        else:
            array.append(right[j])
            j+=1

    array.extend(left[i:])
    array.extend(right[j:])

    return array

if __name__ == "__main__":
    x = [5, 3, 8, 4, 2]
    print(merge_sort(x))



